pkg/util/fxutil/test.go

```diff
@@ -0,0 +1,110 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package fxutil
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/spf13/cobra"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/fx"
+	"go.uber.org/fx/fxtest"
+)
+
+// IsTest is true if this is a test run.  This will always be false in
+// "real" binaries.
+//
+// This can be used to ensure that undesirable behavior such as reporting
+// metrics or accessing on-disk configuration files does not occur during
+// tests.
+var IsTest bool
+
+// Test runs a test case within an fx.App.
+//
+// The given function is called after the app's startup has completed, with its
+// arguments filled via Fx's dependency injection.  fxutil.IsTest is set for the
+// duration of the function call.  Within the app, `t` is provided as type
+// `testing.TB`.
+//
+// Use `fx.Options(..)` to bundle multiple fx.Option values into one.
+func Test(t testing.TB, opts fx.Option, fn interface{}) {
+	IsTest = true
+	defer func() { IsTest = false }()
+
+	delayed := newDelayedFxInvocation(fn)
+	app := fxtest.New(
+		t,
+		fx.Supply(fx.Annotate(t, fx.As(new(testing.TB)))),
+		delayed.option(),
+		opts,
+	)
+	defer app.RequireStart().RequireStop()
+
+	if err := delayed.call(); err != nil {
+		t.Fatal(err.Error())
+	}
+}
+
+// TestOneShotSubcommand is a helper for testing commands implemented with fxutil.OneShot.
+//
+// It takes an array of commands, and attaches all to a temporary top-level
+// command, then executes the given command line (beginning with the
+// subcommand name) against that top-level command.
+//
+// The execution should eventually call fxutil.OneShot with the oneShotFunc
+// given by expectedOneShotFunc.  However, this function will not actually be
+// called, as that would lead to the one-shot command actually running.  It
+// is validated with fx.ValidateApp, however.
+//
+// The `fx.Option`s passed to fxutil.OneShot are used to create a new app
+// containing only the final argument to this function.  Be careful not to
+// require any components, since nothing is mocked here.  Typically, the
+// function only requires static values such as `BundleParams` or `cliParams`
+// and asserts they contain appropriate values.
+func TestOneShotSubcommand(
```

Comment:
> I don't see a lot of value of testing the initialization of cobra.Command

I think others disagree -- we are currently contemplating adding additional QA of precisely this (ASC-201), because changing it is risky.  So, adding testing is a worthwhile pursuit.

The first suggestion works for subcommands with only one *cobra.Command, but there are many subcommands that have further sub-subcommands, and they would need `commands(globalParams, subsubcommand1, subsubcommand2, subsubcommand3, ..)`.  So that ends up being quite a bit of additional boilerplate just to avoid some test-only shared state.

Another option may be to provide a mockable fxutil, passed in from the top-level functions.  So
```go
func main() {
    if err := command.MakeCommand(fxutil.Real(), subcommands.AgentSubcommands()).Execute(); err != nil {
        os.Exit(-1)
    }
}
```
and
```go
func MakeCommand(fxutil fxutil.Interface, subcommandFactories []SubcommandFactory) *cobra.Command {
    ...
    for _, sf := range subcommandFactories {
        subcommands := sf(fxutil, &globalParams)
        for _, cmd := range subcommands {
            agentCmd.AddCommand(cmd)
        }
    }

    return agentCmd
}
```
and
```go
func Commands(fxutil fxutil.Interface, globalParams *command.GlobalParams) []*cobra.Command {
    ...
        return fxutil.OneShot(someFunc,
    ...
```
and in tests
```go
func TestInstallCommand(t *testing.T) {
    fxutil.TestOneShotSubcommand(t,
        func (fxutil fxutil.Interface) []*cobra.Command {
            return Commands(fxutil, &command.GlobalParams{})
        },
        []string{"integration", "install", "foo==1.0", "-v"},
        install,
        ...
```

I'm just not sure that's worthwhile?  I don't think we need to be absolutist about eliminating use of globals.  For the cost of a pointer, the existing solution has the advantage of being invisible outside of the fxutil package: the Commands functions just do what the code says, and the TestFooCommand functions can test this effectively.

Reply:
> I don't see a lot of value of testing the initialization of cobra.Command
> I think others disagree --

Just to make it clear I was speaking about testing the initialization of `cobra.Command` like in [this place](https://github.com/DataDog/datadog-agent/pull/13718/files#diff-ab7b7208bc29d9db76f44a24ac3117866e007144e084fd0aa2e060d43bbf208fR118-R125), but indeed there are other locations like [this one](https://github.com/DataDog/datadog-agent/pull/13722/files#diff-df2ea1bee97aade82c8b92da5d7246c332ff895c23baf4481cc97d8e8ac68054R117-R125) where it makes more sense.

Reply:
If we're concerned about this override global (I'm not super concerned), we could also pass the override function as a param to `fxutil.Oneshot()` (though that'd require modifying the interface) and nil otherwise.